src/components/views/elements/AccessibleButton.js

```diff
@@ -27,8 +29,22 @@ import React from 'react';
 export default function AccessibleButton(props) {
     const {element, onClick, children, ...restProps} = props;
     restProps.onClick = onClick;
+    // We need to consume enter onKeyDown and space onKeyUp
+    // otherwise we are risking also activating other keyboard focusable elements
+    // that might receive focus as a result of the AccessibleButtonClick action
+    restProps.onKeyDown = function(e) {
```

Comment:
I'm confused as to why we would treat SPACE and ENTER differently here. I'd expect them to both be handled `onKeyDown` because if for some bizarre reason pressing space results in focus arriving on a button, we really shouldn't be handling `onKeyUp` for space at that point.

Reply:
I can't explain this very well I was playing with this trying out various combinations of onKeyDown and onKeyUp and what I think is most important is that we stop propagating and prevent default of both onKeyDown and onKeyUp. I am not going to change this unless you can show me an example which is working well in both these cases:

- Go to the Riot settings, use the tab key to navigate to the Logout button, press either enter key or space key and see the dialog. The issue here is that depending on the default browser behaviour the default dialog button receives the focus as the dialog appears and it then gets activated.
- If the logout dialog has not been dismissed then press the tab key until the Cancel button receives the focus. Now test how if you can close the dialog by pressing both the enter key and the space bar key. The issue here is that the logout button in the settings might receive the focus as the dialog is closing and if the proper event is not consumed the dialog might reactivate.

You need to test both these cases with space key and the enter key.
I know this feels very broken however we are only handling the differences imposed by the fact we have various button types i.e. the native button and our own AccessibleButton and the resulting behaviour of their differences.

Reply:
OK, so what's actually going on here is that for a actual HTML `button` (eg. the cancel button in the Sign Out dialog), activating it with enter swallows the keydown event but propagates a keyup event. Activating it with space swallows both, which is why it happens to work in this configuration. This is still a little buggy though because the spacebar press still propagates out when activating the 'sign out' button with the spacebar (the settings page scrolls down as the dialog appears).

This may be an argument for changing everything to consistently use onKeyDown (it was originally changed on onKeyUp by me here: https://github.com/matrix-org/matrix-react-sdk/pull/742#discussion_r104937718). The problem with using keyup for space and keydown for enter will be that we may get this same problem again (ie. the accessiblebutton captures keydown but propagates keyup which triggers something else that it shouldn't do.

I'd probably argue for switching to keydown everywhere since this empirically what a native `button` does, but I'd also like to split this change out of this PR, because the rest of this PR looks great and it's already very large. In future it's probably better to split these changes up into smaller PRs.

Reply:
Please try the following two sets of steps


===== Take 1 =====

* Apply this PR
* Edit AccessibleButton.js file in a way so both space and enter key presses will trigger onKeyDown and remove onKeyUp all together
* Build the resulting app
* Navigate to the Riot settings.
* When the Settings window comes up keep pressing the tab key until Logout button receives the focus
* Press the space bar key


====== Results ======

Damn! We were expecting a confirmation dialog however we have just logged out because of the underlying issue :(


====== Expected results ======

Before logging out we should be able to confirm our choice within the question dialog.


===== Take 2 =====

* Apply this PR
* Edit AccessibleButton.js file in such a way that both enter and space bar keypresses will trigger onKeyDown i.e. onClick will be executed when either of them is pressed onKeyDown.
* Keep onKeyUp method present however change it so it will do nothing more except of e.preventDefault() and e.stopPropagation()
* Continue with the steps as if you were following take 1.


====== Results ======

Damn! We were expecting a confirmation dialog however we have just logged out because of the underlying issue :(


====== Expected results ======

Before logging out we should be able to confirm our choice within the question dialog.


So I can't explain it very well however by some tryal and error this is the only walid implementation as far as I can test.

Reply:
Yep, I suspect this is because we're using keydown/keyup inconsistently in some other file, hence suggesting we should use one of them consistently throughout the app.

Reply:
@dbkr isn't the point that browsers don't do onKeyUp vs. onKeyDown consistently the actual problem here? And by mimicking the weirdness as in the PR, we get something that works but that feels wrong.

Reply:
I don't know - if this is the case I think we should find out what this inconsistency is and document it in a comment in this block of code, otherwise it's going to continue to confuse people.

Reply:
Will it help if I change that inline comments to say something like this?

 	    // We need to consume enter onKeyDown and space onKeyUp
    // otherwise we are risking also activating other keyboard focusable elements
    // that might receive focus as a result of the AccessibleButtonClick action
    // It's because we are using html buttons at a few places e.g. inside dialogs
    // And divs which we report as role button to assistive technologies.
    // Browsers handle space and enter keypresses differently and we are only adjusting to the
    // inconsistencies here

Reply:
OK, it at least gives a hint as to what's going on. Can we fix the fact that the spacebar keydown still leaks out when activating an accessible button with the spacebar though?